contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,220 @@
+    }
+
+    /**
+     * @dev Grants `role` to `account`.
+     *
+     * If `account` had not been already granted `role`, emits a {RoleGranted}
+     * event.
+     *
+     * Requirements:
+     *
+     * - the caller must have ``role``'s admin role.
+     *
+     * May emit a {RoleGranted} event.
+     */
+    function grantRole(bytes32 context, bytes32 role, address account) public virtual onlyRole(context, getRoleAdmin(role)) returns (bool) {
+        return _grantRole(context, role, account);
+    }
+
+    /**
+     * @dev Revokes `role` from `account`.
+     *
+     * If `account` had been granted `role`, emits a {RoleRevoked} event.
+     *
+     * Requirements:
+     *
+     * - the caller must have ``role``'s admin role.
+     *
+     * May emit a {RoleRevoked} event.
+     */
+    function revokeRole(bytes32 context, bytes32 role, address account) public virtual onlyRole(context, getRoleAdmin(role)) {
+        _revokeRole(context, role, account);
+    }
+
+    /**
+     * @dev Revokes `role` from the calling account.
+     *
+     * Roles are often managed via {grantRole} and {revokeRole}: this function's
+     * purpose is to provide a mechanism for accounts to lose their privileges
+     * if they are compromised (such as when a trusted device is misplaced).
+     *
+     * If the calling account had been revoked `role`, emits a {RoleRevoked}
+     * event.
+     *
+     * Requirements:
+     *
+     * - the caller must be `callerConfirmation`.
+     *
+     * May emit a {RoleRevoked} event.
+     */
+    function renounceRole(bytes32 context, bytes32 role, address callerConfirmation) public virtual {
+        if (callerConfirmation != _msgSender()) {
+            revert EnhancedAccessControlBadConfirmation();
+        }
+
+        _revokeRole(context, role, callerConfirmation);
+    }
+
+    /**
+     * @dev Sets `adminRole` as ``role``'s admin role.
+     *
+     * Emits a {RoleAdminChanged} event.
+     */
+    function _setRoleAdmin(bytes32 role, bytes32 adminRole) internal virtual {
+        bytes32 previousAdminRole = getRoleAdmin(role);
+        _adminRoles[role] = adminRole;
+        emit EnhancedAccessControlRoleAdminChanged(role, previousAdminRole, adminRole);
+    }
+
+    /**
+     * @dev Revoke all assignments of a given role in a given context.
+     */
+    function _revokeRoleAssignments(bytes32 context, bytes32 role) internal virtual {
+        _roleVersion[context][role]++;
+    }
+
+    /**
+     * @dev Attempts to grant `role` to `account` and returns a boolean indicating if `role` was granted.
+     *
+     * Internal function without access restriction.
+     *
+     * May emit a {RoleGranted} event.
+     */
+    function _grantRole(bytes32 context, bytes32 role, address account) internal virtual returns (bool) {
+        if (!hasRole(context, role, account)) {
+            _roles[context][role][account].hasRole = true;
+            _roles[context][role][account].version = _roleVersion[context][role];
```

Comment:
Worth verifying, but I believe that both variables will be packed into a single slot, and a setter using a struct literal will thus be more efficient.